File: hotseat/inspect.py
"""What a stopped piece of work was actually doing.

A list of stopped agents is only half useful: deciding whether to continue one
means knowing what it was in the middle of. This recovers that from whatever the
two sources kept.

Everything here is read-only and local. Content is truncated on the way out rather
than in the caller, so a long transcript cannot flood a terminal, and speech markup
is stripped because it is delivery scaffolding, not what the agent said.
"""

from __future__ import annotations

import json
from pathlib import Path
import re
import sqlite3

from .paths import PROJECTS_DIR
from .textutil import clean as _clean, entry_text, trim

#: A continuation prompt. It is not something the user typed, but its summary is
#: often the best description of what the work was, so it is mined rather than
#: discarded.
CONTEXT_PREAMBLE = "This session is being continued from a previous conversation"
SUMMARY_MARKER = re.compile(r"(?:Summary:|Primary Request and Intent:?)\s*(.+)", re.S)


class InspectError(RuntimeError):
    """The stopped work could not be inspected."""


def _summary_of(text: str | None) -> str:
    """Pull the work description out of a context-continuation preamble."""
    if not text or not text.startswith(CONTEXT_PREAMBLE):
        return ""
    match = SUMMARY_MARKER.search(text)
    return trim(match.group(1) if match else text[len(CONTEXT_PREAMBLE):], 500)





# --- native ---------------------------------------------------------------

def _native_detail(session_id: str) -> dict | None:
    matches = list(PROJECTS_DIR.glob(f"*/{session_id}.jsonl"))
    if not matches:
        # Accept the short form shown in listings.
        matches = list(PROJECTS_DIR.glob(f"*/{session_id}*.jsonl"))
    if not matches:
        return None
    transcript = matches[0]

    entries = []
    try:
        with transcript.open(errors="replace") as handle:
            for line in handle:
                try:
                    entries.append(json.loads(line))
                except ValueError:
                    continue
    except OSError as exc:
        raise InspectError(f"could not read the transcript: {exc}") from exc
    if not entries:
        return None

    def usable(entry, role):
        if entry.get("type") != role or entry.get("isApiErrorMessage"):
            return False
        text = entry_text(entry).strip()
        # Tool results and system scaffolding arrive as user entries too.
        return bool(text) and not text.startswith(("<", CONTEXT_PREAMBLE))

    users = [e for e in entries if usable(e, "user")]
    summary = next((_summary_of(entry_text(e)) for e in reversed(entries)
                    if _summary_of(entry_text(e))), "")
    assistants = [e for e in entries if usable(e, "assistant")]
    last = entries[-1]

    return {
        "kind": "native",
        "id": transcript.stem,
        "name": transcript.stem[:8],
        "backend": "claude",
        "model": next((e.get("message", {}).get("model") for e in reversed(entries)
                       if (e.get("message") or {}).get("model")), None),
        "cwd": last.get("cwd") or next((e.get("cwd") for e in reversed(entries)
                                        if e.get("cwd")), None),
        "branch": next((e.get("gitBranch") for e in reversed(entries)
                        if e.get("gitBranch")), None),
        "last_user": trim(entry_text(users[-1])) if users else "",
        "summary": summary,
        "last_assistant": trim(entry_text(assistants[-1])) if assistants else "",
        "last_activity": last.get("timestamp"),
        "queued": [],
        "recent": [{"role": e.get("type"), "text": trim(entry_text(e), 160)}
                   for e in entries if usable(e, "user") or usable(e, "assistant")][-6:],
        "turns": len(users),
    }


def detail(identifier: str) -> dict:
    """Everything known about one stopped item, by session slug or transcript id."""
    from .plugins import inspect_detail
    found = inspect_detail(identifier)
    if found is None:
        found = _native_detail(identifier)
    if found is None:
        raise InspectError(f"nothing found for {identifier!r}")
    return found

File: internal/clarp/inspect.go
package clarp

// What a stopped Clarp agent was actually doing.
//
// A list of stopped agents is only half useful: deciding whether to continue one
// means knowing what it was in the middle of. This recovers that from what Clarp
// kept: its message log and its turn queue. Content is truncated on the way out
// rather than in the caller, so a long transcript cannot flood a terminal, and
// speech markup is stripped because it is delivery scaffolding, not what the
// agent said.

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strings"

	"github.com/Maxteabag/hotseat/internal/work"
)

// QueuedTurn is a prompt waiting in a Clarp agent's queue.
type QueuedTurn struct {
	Origin *string `json:"origin"`
	Text   string  `json:"text"`
}

// Inspect is everything known about one Clarp agent, by session slug. It
// returns nil, nil when no state database exists or no live agent has that
// session, so a caller can fall through to work.Inspect for native transcripts.
// That is the merge order the Python plugin used: Clarp answers first, then the
// native transcript, then "nothing found".
func (s *Service) Inspect(ctx context.Context, identifier string) (*work.Detail, error) {
	if !stateExists(s.StatePath) {
		return nil, nil
	}
	detail, err := s.clarpDetail(ctx, identifier)
	if err != nil {
		return nil, &work.InspectError{Msg: fmt.Sprintf("could not read Clarp state: %v", err)}
	}
	return detail, nil
}

type message struct {
	role, text, timestamp sql.NullString
}

func (s *Service) clarpDetail(ctx context.Context, session string) (*work.Detail, error) {
	db, err := s.open()
	if err != nil {
		return nil, err
	}
	defer db.Close()

	var agentID string
	var persona, backend, cwd, model sql.NullString
	err = db.QueryRowContext(ctx,
		`SELECT agent_id, persona, backend, cwd, model FROM agents
                    WHERE session = ? AND deleted_at IS NULL`, session).
		Scan(&agentID, &persona, &backend, &cwd, &model)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	rows, err := db.QueryContext(ctx,
		`SELECT role, text, timestamp FROM messages
                    WHERE agent_id = ? AND role IN ('user','assistant')
                      AND length(trim(coalesce(text,''))) > 0
                    ORDER BY seq DESC LIMIT 200`, agentID)
	if err != nil {
		return nil, err
	}
	var newestFirst []message
	for rows.Next() {
		var m message
		if err := rows.Scan(&m.role, &m.text, &m.timestamp); err != nil {
			rows.Close()
			return nil, err
		}
		newestFirst = append(newestFirst, m)
	}
	rows.Close()
	if err := rows.Err(); err != nil {
		return nil, err
	}

	queuedRows, err := db.QueryContext(ctx,
		`SELECT text, origin, enqueued_at FROM queued_turns
                    WHERE agent_id = ? AND status = 'queued'
                    ORDER BY queue_seq`, agentID)
	if err != nil {
		return nil, err
	}
	queued := []any{}
	for queuedRows.Next() {
		var text, origin sql.NullString
		var enqueued sql.NullInt64
		if err := queuedRows.Scan(&text, &origin, &enqueued); err != nil {
			queuedRows.Close()
			return nil, err
		}
		queued = append(queued, QueuedTurn{Origin: nullable(origin), Text: work.Trim(text.String, 300)})
	}
	queuedRows.Close()
	if err := queuedRows.Err(); err != nil {
		return nil, err
	}

	// newestFirst is the query order; the Python reversed it into "ordered" and
	// then walked that backwards, which is the same thing.
	detail := &work.Detail{
		Kind:    "clarp",
		ID:      session,
		Name:    persona.String,
		Backend: backend.String,
		Model:   nullable(model),
		CWD:     nullable(cwd),
		Queued:  queued,
		Recent:  []work.RecentTurn{},
	}
	if detail.Name == "" {
		detail.Name = session
	}
	for _, m := range newestFirst {
		if m.role.String != "user" {
			continue
		}
		if !strings.HasPrefix(m.text.String, work.ContextPreamble) {
			detail.LastUser = work.Trim(m.text.String, work.Snippet)
			break
		}
	}
	for _, m := range newestFirst {
		if m.role.String != "user" {
			continue
		}
		if summary := work.SummaryOf(m.text.String); summary != "" {
			detail.Summary = summary
			break
		}
	}
	for _, m := range newestFirst {
		if m.role.String == "assistant" {
			detail.LastAssistant = work.Trim(m.text.String, work.Snippet)
			break
		}
	}
	if len(newestFirst) > 0 {
		detail.LastActivity = nullable(newestFirst[0].timestamp)
	}
	recent := newestFirst
	if len(recent) > 6 {
		recent = recent[:6]
	}
	for i := len(recent) - 1; i >= 0; i-- {
		detail.Recent = append(detail.Recent, work.RecentTurn{
			Role: recent[i].role.String,
			Text: work.Trim(recent[i].text.String, 160),
		})
	}
	return detail, nil
}

// nullable is a NULL-able column as the pointer the JSON shape wants: nil for
// NULL, the value otherwise, even when empty.
func nullable(value sql.NullString) *string {
	if !value.Valid {
		return nil
	}
	s := value.String
	return &s
}
